BQN updates

AndersonTorres opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

AndersonTorres committed:
dbqn: init at 0.0.0+unstable=2021-10-02

Many thanks @sternenseemann for the expression. I made a rewrite to keep it
clean.

sternenseemann commented:
Let's use the original PRs #140268 and #139572. I don't think recommitting these unrelated changes squashed together does any good. I also don't agree with the changes you've made, but in this form it is hard to review them.

SuperSandro2000 commented on pkgs/development/interpreters/bqn/dzaima-bqn/default.nix:
```diff
@@ -0,0 +1,72 @@
+    mkdir -p output
+    env LC_ALL=C.UTF-8 javac --release 8 -d ./output $(find src -name '*.java')
+    (cd output; jar cvfe ../BQN.jar BQN.Main *)
+    rm -fr output
+
+  '' + lib.optionalString buildNativeImage ''
+    native-image --report-unsupported-elements-at-runtime -jar BQN.jar dbqn
+  '' + ''
+
+    runHook postBuild
+  '';
```
```suggestion
    rm -r output
  '' + lib.optionalString buildNativeImage ''
    native-image --report-unsupported-elements-at-runtime -jar BQN.jar dbqn
  '' + ''
    runHook postBuild
  '';
```

SuperSandro2000 commented on pkgs/development/interpreters/bqn/dzaima-bqn/default.nix:
```diff
@@ -0,0 +1,72 @@
+  };
+
+  nativeBuildInputs = [
+    jdk
```
If it is used in makeWrapper it does not need to be added here and if used at runtime it should be buildInputs.

AndersonTorres replied:
`native-image` is not shown in PATH if it is not added to this list.

SuperSandro2000 commented on pkgs/development/interpreters/bqn/dzaima-bqn/default.nix:
```diff
@@ -0,0 +1,72 @@
+
+    makeWrapper "${lib.getBin jdk}/bin/java" "$out/bin/dbqn" \
+      --add-flags "-jar $out/lib/dbqn/BQN.jar"
+  '') + ''
+
+    ln -s $out/bin/dbqn $out/bin/bqn
```
```suggestion
  '') + ''
    ln -s $out/bin/dbqn $out/bin/bqn
```

SuperSandro2000 commented on pkgs/development/interpreters/bqn/dzaima-bqn/default.nix:
```diff
@@ -0,0 +1,72 @@
+
+  meta = with lib; {
+    homepage = "https://github.com/dzaima/BQN";
+    description = "A BQN implementation in Java"
+                  + lib.optionalString buildNativeImage ", compiled as a GraalVM native image";
```
```suggestion
    description = "A BQN implementation in Java" + lib.optionalString buildNativeImage ", compiled as a GraalVM native image";
```

AndersonTorres committed:
mbqn: init at 0.0.0+unstable=2021-10-01

AndersonTorres committed:
cbqn: 0.0.0+unstable=2021-09-29 -> 0.0.0+unstable=2021-10-01

Many thanks for the various rewritings by @sternenseemann!

- The intermediary recompilation phases are hidden by default inside a
  `lib.dontRecurseIntoAttrs`; only the zeroth and the second ones are exposed.

- `clangStdenv` is used by default in non-cross builds.

- The standalone build has an inferior priority.

- The mbqn sources are explicitly taken from mbqn instead of hardcoded in cbqn.

sternenseemann commented:
Why have you self-merged this without even replying to my expressed concerns?

AndersonTorres commented:
I have looked at the review given by Sandro, and adjusted the code accordingly.
It is easier to review a monolytic squash instead of many small stacked commits, even more when they interact with each other.

Given that no one opposed to anything after that, I made the last adjustements and merged it.

sternenseemann commented:
> I also don't agree with the changes you've made, but in this form it is hard to review them.

https://github.com/NixOS/nixpkgs/pull/140319#issuecomment-932827176

Would say that counts as an objection…
